File: src/main/java/kr/hhplus/be/server/infrastructure/RedisLockManager.java
package kr.hhplus.be.server.infrastructure;

import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Component;

import java.time.Duration;

@Component
public class RedisLockManager {

    private final RedisTemplate<String, Object> redisTemplate;

    public RedisLockManager(RedisTemplate<String, Object> redisTemplate) {
        this.redisTemplate = redisTemplate;
    }

    /**
     * 단순 SETNX + TTL 기반 락 획득
     * 소유자 정보(Token)를 검증하지 않는 단순한 구현 (아쉬운 사례 재현)
     */
    public boolean tryLock(String key, long timeoutSeconds) {
        // value는 단순히 "lock" 상수로 고정
        return Boolean.TRUE.equals(redisTemplate.opsForValue()
                .setIfAbsent(key, "lock", Duration.ofSeconds(timeoutSeconds)));
    }

    /**
     * 단순 DEL 기반 락 해제
     * 소유자 확인 없이 무조건 삭제 (아쉬운 사례 재현)
     */
    public void unlock(String key) {
        // TTL 만료 후 다른 요청이 락을 획득했을 때, 이전 요청이 unlock을 호출하면 남의 락을 해제할 위험이 있음
        redisTemplate.delete(key);
    }
}

File: src/main/kotlin/kr/hhplus/be/server/infrastructure/RedisLockManager.kt
package kr.hhplus.be.server.infrastructure

import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.stereotype.Component
import java.util.concurrent.TimeUnit

/**
 * Redis 기반 분산락 매니저
 *
 * 단순한 SET NX + TTL 방식으로 구현했습니다.
 * - 락 소유자 식별 없음
 * - TTL 만료 시나리오에 대한 방어 없음
 *
 * 이는 교육용 PR의 일부로, 실제 운영 환경에서는 추가적인 안전장치가 필요합니다.
 */
@Component
class RedisLockManager(
    private val redisTemplate: StringRedisTemplate,
) {
    /**
     * 분산락 획득 시도
     *
     * @param key 락 키
     * @param timeoutMs 락 유지 시간 (밀리초)
     * @return 락 획득 성공 여부
     */
    fun tryLock(key: String, timeoutMs: Long): Boolean {
        val result =
            redisTemplate.opsForValue()
                .setIfAbsent(key, "lock", timeoutMs, TimeUnit.MILLISECONDS)
        return java.lang.Boolean.TRUE == result
    }

    /**
     * 분산락 해제
     *
     * 주의: 현재 구현은 락 소유자 검증 없이 삭제합니다.
     * TTL 만료 후 다른 요청이 락을 획득한 경우에도 기존 요청이 unlock 할 수 있습니다.
     */
    fun unlock(key: String) {
        redisTemplate.delete(key)
    }
}
